strapi/strapi
# [V4] Add support to have multiple Strapi instances in same database using table prefix

### What does it do?

Add support to have multiple Strapi instances use the same table by making use of table prefixes.

- `examples/getstarted/config/database.js`
Lets user specify table prefix to use for the particular instance.
- `packages/core/database/lib/dialects/[sqlite|mysql|postgres]/schema-inspector.js`
Update `LIST_TABLE` SQL-query to only return the tables associated with the particular Strapi instance (i.e. only return tables that have the prefix).
- `packages/core/database/lib/index.js`
Add field `tablePrefix` to database class and pass it to `createMetadata` to make all tables to be created gets the correct prefix.
- `packages/core/database/lib/metadata/index.js`
Prepend prefix to table name.
- `packages/core/database/lib/migrations/index.js` and `packages/core/database/lib/migrations/storage.js`
Add table prefix to migration table.
- `packages/core/database/lib/schema/diff.js`
Extend checks that are used to determine whether a table has a reserved name or not to handle table prefixes.
- `packages/core/database/lib/schema/storage.js`
Add prefix support for the `strapi_database_schema`-table.
- `packages/core/database/lib/utils/table-prefix.js`
Function to validate and create the table prefix. 

### Why is it needed?

This PR solves the problem of using the same database for multiple Strapi instances. Typically some hosting providers have database limits and by using prefixes the same database can be reused for multiple purposes (e.g. multiple Strapi instances).

### How to test it?

Configure two Strapi-instances to use different table prefixes but make them communicate to the same database. In the case of SQLite, you create a `data.db` file and then set the same path for both instances in `examples/getstarted/config/database.js`. The same applies for MySQL and Postgres.

The easiest way to create the two instances is to simply duplicate the examples/getstarted folder and make sure to use different server port (`config/server.js`) as well as a different database prefix in `config/database.js`.

### Related issue(s)/PR(s)

fixes: #5080 
Related PR: #11491 (Was closed when the `releases/v4` branch was deleted)

### State of PR
This PR is ready to be merged.
